### ISSUE TITLE
[MDEPLOY-318] version 3.1.2 is uploading jar files to wrong endpoints

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-deploy-plugin#203]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** opened **[MDEPLOY-318](https://issues.apache.org/jira/browse/MDEPLOY-318?redirect=false)** and commented

hi, i'm using spring-boot, and upon updating to 3.2.6 i noticed that my parent pom was uploaded corrupted.

`[FATAL] Non-parseable POM ~/.m2/repository/nl/intractief/intr-parent/7.1.5/intr-parent-7.1.5.pom: only whitespace content allowed before start tag and not P (position: START_DOCUMENT seen P... @1:2)  @ ~/.m2/repository/nl/intractief/intr-parent/7.1.5/intr-parent-7.1.5.pom, line 1, column 2`

 
* the same parent pom was correctly published using spring-boot 3.2.5
* the parent pom is correct before publishing in the pipeline (manually checked the target folder).
* after publishing i see binary stuff as pom (guess: zipped binary content)
* setting the maven-deploy-plugin.version in the project to 3.1.1 solves the issue.

in the spring-boot project i created an issue about this (see external issue link) and found another similar issue with a functional workaround.

we think there is an issue with the deployment endpoints in version 3.1.2 it seems the jars end up at the pom endpoints

 

 

---

**Affects:** 3.1.2

**Attachments:**
- [pom.xml](https://issues.apache.org/jira/secure/attachment/13069276/pom.xml) (_1.19 kB_)
- [test-maven-deploy-0.0.1.zip](https://issues.apache.org/jira/secure/attachment/13069277/test-maven-deploy-0.0.1.zip) (_1.77 kB_)
- [test-maven-deploy-0.0.2.zip](https://issues.apache.org/jira/secure/attachment/13069278/test-maven-deploy-0.0.2.zip) (_1.77 kB_)
- [test-maven-parent-0.0.1.pom](https://issues.apache.org/jira/secure/attachment/13069279/test-maven-parent-0.0.1.pom) (_1.64 kB_)
- [test-maven-parent-0.0.2.pom](https://issues.apache.org/jira/secure/attachment/13069280/test-maven-parent-0.0.2.pom) (_1.18 kB_)

**Issue Links:**
- [MNG-8138](https://issues.apache.org/jira/browse/MNG-8138) Maven internal state should not allow to become "broken"

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Can you provide some minimal reproducer?

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Seems strange that for you m-deploy-p 3.1.1 works while 3.1.2 does not, as literally the release notes is this: https://issues.apache.org/jira/secure/ReleaseNote.jspa?projectId=12317228&version=12354128

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

i know, it's patch updates but the deploy really works when downgrading the deploy plugin.

i deploy from my gitlab-ci pipeline using

`mvn jar:jar deploy:deploy -pl $MAVEN_MODULES -DaltDeploymentRepository=gitlab-maven::$MAVEN_REPOSITORY`

but it only displays incorrect artifacts in the repository, so minimal reproduction requires a maven repository.

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

https://github.com/spring-projects/spring-boot/issues/40919 also contains some http debug log of the requests

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Sure, just create a reproducer with some (even your URL to some repo) and I will replace it with some I have write access to.

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

working on it in https://gitlab.com/tub/test-maven-deploy/ that includes the minimal pipeline i use

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

ok
* https://gitlab.com/tub/test-maven-deploy/-/releases/0.0.1 this is the faulty deployed version
* https://gitlab.com/tub/test-maven-deploy/-/releases/0.0.2 this is the correct one

do you want me to add the source? (the above repo is public)

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

[^pom.xml]

this is the pom; so the url should differ and you need credentials to publish, i change the used version using the maven-deploy-plugin.version property

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

faulty version

[^test-maven-deploy-0.0.1.zip][^test-maven-parent-0.0.1.pom]

correct version

[^test-maven-deploy-0.0.2.zip][^test-maven-parent-0.0.2.pom]

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

i think it's something here, i think determinating Artifact's deserves its own function and test

https://github.com/apache/maven-deploy-plugin/pull/47/files#diff-ffcc9777ab63d54259e77f8e53ab3fc9c1b4ae1de9554187ab2a097b41b9bd76

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Unsure how to run this thing? Also, the pom is packaging "pom" so from where comes the ZIP? When I invoke this (for example `{}mvn package{`}) nothing happens, as expected in a way, as packaging is "pom"...

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

So, when I do deploy this happens: https://gist.github.com/cstamas/6466d942dbd56aa96e341432315b846d

Again, unsure from where the ZIP should come in this reproducer

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

With 3.1.2 of m-deploy-p same thing: https://gist.github.com/cstamas/a6fa1997c4922981dbced143127dce78

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

As I see, your problem is not Maven or Deploy plugin 3.1.2 but your CI script: https://gitlab.com/tub/test-maven-deploy/-/blob/main/.gitlab-ci.yml

That script is totally off: you explicitly invoke `jar:jar` and `deploy:deploy` 

Please look at the build lifecycles instead: https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

k, tnx, will look into that, i think we invoke the specific mojo's to prevent re-building the whole project when using the lifecycle (because of the split-up into script and after_script in the .gitlab-ci

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

if i omit the jar:jar in that gitlab-ci after_script it gets published correctly

it's there because in the full script i also publish jars from this project (it's multi-module and the parent pom is one of the modules)

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

thank you! :)

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

You welcome :)

If you have any other issue, please feel free to report them. Am interested in this issue of your "we invoke the specific mojo's to prevent re-building the whole project"... can you elaborate that more?

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

the gitlab-ci file in the project is good example of that

i like to re-use my build scripts as much as possible; in this case that entails the main script part to be the same for the normal build (every commit) and the tag build (for releases)

so for a release i set all versions to the same as the git tag in the pre_script using

mvn $MAVEN_CLI_OPTS versions:set -DprocessAllModules -DnewVersion=$CI_COMMIT_TAG

and in the after_script i invoke the deploy:deploy on the modules i'd like to deploy (normally that after_script is only for tag builds and specifies the modules using -pl )

mvn $MAVEN_CLI_OPTS jar:jar deploy:deploy -pl $MAVEN_MODULES -DaltDeploymentRepository=$MAVEN_SERVER_ID::$MAVEN_REPOSITORY

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Created MNG-8137 as follow up, as something is fishy here.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Created MNG-8138 as yet another follow up.

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

in a multi-module project i need to trigger jar-type modules different than pom type modules

e.g. two commands

mvn deploy:deploy -pl parent-pom
mvn jar:jar deploy:deploy -pl some-client

if i do deploy:deploy on packaging jar type modules without jar:jar i get:

`Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:3.1.1:deploy (default-cli) on project some-client: The packaging for this project did not assign a file to the build artifact -> [Help 1]`

i think 3.1.2 will behave the same here.`{}{`}

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

no the reason i'm using the deploy:deploy in the first place is that i dont want all modules from this project te be deployed; some of them (spring-boot server modules) are processed further down the pipeline to docker images

but currently i have multiple workarounds for this issue, i think for you guys it's better to focus on the better solution in MNG-8137 or MNG-8138

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

See https://maven.apache.org/guides/introduction/introduction-to-profiles#property (at very end of section)

**Comment by jira-importer** [comment]
**[Juan Antonio Brena Moral](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER307772)** commented

Hi,

I noticed that the ticket continue opened and I would like to report than 2 weeks ago using the latest maven plugin 3.1.3 included in {_}Spring Boot 3.3.4{_}, it is not working for us when we try to deploy to JFROG Artifactory or Azure Packages with Azure Pipelines. 

The workaround that we had to do in order to deploy was to downgrade the plugin to `maven-deploy-plugin:3.1.1:`

 

`          - task: Maven@4`
`            displayName: Deploy jar`
`            inputs:`
`              mavenPomFile: 'pom.xml'`
`              goals: 'jar:jar org.apache.maven.plugins:maven-deploy-plugin:3.1.1:deploy'`
`              ...`

 

I would like to collaborate to evolve the plugin and help in the testing.

Juan Antonio

**Comment by jira-importer** [comment]
**[Michaël](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER305673)** commented

The root cause of this issue is discussed in https://issues.apache.org/jira/browse/MNG-8138

TLDR; invoking the jar:jar task directly changes the (internal) state of a pom project/module to think it has a jar to deploy. The deploy plugin consequently tries to honor that and that is the behaviour you are observing.

The "best" way of solving this is not invoking the tasks directly but using the maven lifecycles and configuring your project accordingly.

A shortcut worth trying is to mark skip the maven-jar-plugin for the pom module (mentioned in above ticket)

`<plugin> {color`}

`    <groupId>org.apache.maven.plugins</groupId> {color`}

`     <artifactId>maven-jar-plugin</artifactId>{color`}

` <configuration> <skipIfEmpty>true</skipIfEmpty> </configuration>{color`}

`</plugin>`

**Comment by tubbynl** [comment]
link to my original issue in the spring-boot project https://github.com/spring-projects/spring-boot/issues/40952

**Comment by elharo** [comment]
This is not a bug in the `maven-deploy-plugin` that requires a fix in the plugin itself. It is a configuration and lifecycle invocation issue on the user's end, exposing an underlying behavior in core Maven state management (which is tracked separately).

### Root Cause

The user’s CI script manually invokes `jar:jar deploy:deploy` on the command line rather than using the standard Maven lifecycle (e.g., `mvn deploy`).

When `jar:jar` is executed directly from the CLI against a module with `<packaging>pom</packaging>`, it mutates Maven's internal session state, effectively tricking the project into believing it has a JAR artifact attached to it. When `deploy:deploy` executes subsequently in the same session, it attempts to deploy this non-existent or misaligned JAR artifact to the repository endpoints expected for the POM, resulting in corrupted uploads (ZIP data at `.pom` endpoints).

The user explicitly verified that downgrading the deploy plugin to `3.1.1` bypassed the issue, but this was due to tighter artifact checks added in `3.1.2` rather than a regression in deployment logic.

### Resolution & Next Steps

The consensus from Tamas Cservenak and the original author is that the user's CI pipeline was incorrectly bypassing standard Maven lifecycles.

1. **Deploy Plugin:** No action is needed here. The plugin is correctly deploying the artifact state it is handed by Maven core.
2. **Maven Core:** The underlying issue—that Maven allows its internal artifact state to become corrupted/broken when a user directly invokes a Mojo that contradicts the packaging type—is a valid architectural concern. It was spun out into **MNG-8138** (`Maven internal state should not allow to become "broken"`).
3. **User Mitigation:** The user resolved their build by correctly utilizing the `deploy` phase of the default lifecycle, or using `skipIfEmpty` configuration on the `maven-jar-plugin` for POM modules.

You do not need to address this in `maven-deploy-plugin`. Work on preventing this state corruption is best directed at MNG-8138 in Maven core.
